Enable project creation and update via the API

cpeel opened the pull request:
This enables creating and updating projects through the API. Virtually all of the heavy lifting was done in https://github.com/DistributedProofreaders/dproofreaders/pull/754, this just exposes it via the API.

The [api-update-project](https://www.pgdp.org/~cpeel/c.branch/api-update-project/) sandbox has the live code. This is testable via `curl` (as covered in the `USERS_GUIDE.md`) and also via the [Swagger Editor](https://editor.swagger.io/) (be sure to upload the new `dp-openapi.yml` file) via the `https://www.pgdp.org/api-dev/v1` development endpoint.

cpeel commented:
I messed up resolving the conflicts in `USERS_GUIDE.md` from https://github.com/DistributedProofreaders/dproofreaders/pull/786, fixed now.

srjfoo commented:
In my initial testing (via swagger) I found a couple of validation issues:
* If you specify the character suite incorrectly (or use an empty string for the character suite), there's an error response, but the project is created, anyway, since the character suite is the last bit of metadata input.
* The image_source field expects a value from `image_sources.code_name`. I was testing with a `display_name`, which was definitely longer than the 10 characters allowed for the `code_name`. I got a database error from that one.

Off to do some more testing 😁

cpeel commented:
Thanks for finding those, Sharon. I found some other validation corner cases and opened #790 to address them separately from this. It would probably be best if we got that one in first, then I can rebase this branch to pick up those changes.

cpeel commented:
This branch has been refreshed with the improved validation and is ready for more testing.

cpeel commented on api/dp-openapi.yaml:
```diff
@@ -96,6 +96,34 @@ paths:
+            schema:
+              $ref: '#/components/schemas/project'
+      responses:
+        200:
```
For reasons I don't understand, this is what I did in all the other places and I was being consistent in the inconsistency. I probably need a follow-up PR to update all of these globally.

cpeel commented on api/v1_projects.inc:
```diff
@@ -116,37 +116,131 @@ function api_v1_projects($method, $data, $query_params)
+    $project->save();
+
+    if (isset($updates["character_suites"])) {
+        $project->set_charsuites($updates["character_suites"]);
```
This is an oddity in how character suites are associated with projects. Character suites are linked to projects via a linking table, not stored in the `project` table directly. Because of this, we can't set a project's character suites until a project has been created via `save()`. This also means that `Project->validate()` doesn't validate character suites because it is only interested in values for the projects table.

I considered splitting out character suites into a separate endpoint, but projects shouldn't be created without a character suite.

This seemed like the least bad solution to this mess. I'm open to suggestions!

srjfoo commented:
Still poking around at testing (both via Swagger and via `curl`) -- sorry to be so slow about it, though.

Some comments about the documentation, because I'm not sure how much of my struggles are due to my ignorance of things that someone who's actually building a script based on the API should be expected to know.

1. Probably relevant only to testers: I discovered when looking for it that Swagger apparently doesn't understand pagination, nor give access to more than the first page of results that are limited by the API to a default page size.
2. When checking out pagination via `curl` I was confused by the URLs for the pages for a couple of things:
    * It's not clear to me from the documentation that the `; rel=*` at the end of the URL should not be included when building the request for the next page of the results (so I got all sorts of errors when trying to figure that out 😁 ).
    * Once I figured out where to look for the links, I was actually looking at results that had only two pages. Both links pointed to `page=1`, with the `; rel=*` being different. It wasn't until I looked at the pagination links for a request that returned more than 2 pages of results that I realized that the second URL really should, indeed, have pointed to page 2.

cpeel commented:
Good catch on the page number bug - found and fixed that one.

I've updated the User's Guide with a bit more details on the pagination, probably most helpfully pointing to the excellent GitHub docs that describe how to use it. I've also added some details in the Swagger file as well. Frustratingly, while the schema [clearly defines the response headers](https://github.com/cpeel/dproofreaders/blob/api-update-project/api/dp-openapi.yaml#L98) in the required format, the Swagger UI isn't showing the values to the user on a successful query.

srjfoo commented:
Back to testing what I can in Swagger, in `/projects/imagesources`, the following error message:

> can't parse JSON.  Raw result:
>
> \<br />
> <b>Notice</b>:  Undefined index: credit in <b>/home/cpeel/public_html/c.branch/api-update-project/api/v1_projects.inc</b> on line <b>534</b><br />

Looks like something we didn't catch in what's already merged, because I just noticed the following on a project page in the main TEST sandbox:
> Notice: Undefined index: credit in /data/htdocs/c/pinc/Project.inc on line 548

srjfoo commented:
> Back to testing what I can in Swagger, in `/projects/imagesources`, the following error message:
> 
> > can't parse JSON.  Raw result:
> > <br />
> > Notice:  Undefined index: credit in /home/cpeel/public_html/c.branch/api-update-project/api/v1_projects.inc on line 534
> 
> Looks like something we didn't catch in what's already merged, because I just noticed the following on a project page in the main TEST sandbox:
> 
> > Notice: Undefined index: credit in /data/htdocs/c/pinc/Project.inc on line 548

A bit more information: I did some additional testing today, and discovered that there are actually project pages that don't show the second message. Upon investigation, it only occurs when the image source is `_internal`.

cpeel commented:
I missed including the `credits` field for the `_internal` image sources. Since that was in master and will get pushed out to PROD on next week's deployment, I just updated it directly in master in https://github.com/DistributedProofreaders/dproofreaders/commit/71e38476df4a542635eebaaaea4f53e574ac2ecb and rebased this branch.

srjfoo commented:
> I missed including the `credits` field for the `_internal` image sources. Since that was in master and will get pushed out to PROD on next week's deployment, I just updated it directly in master in [71e3847](https://github.com/DistributedProofreaders/dproofreaders/commit/71e38476df4a542635eebaaaea4f53e574ac2ecb) and rebased this branch.

I got that rolled out to TEST, and it fixed the project pages, but I'm still getting the
> <b>Notice</b>:  Undefined index: credit in <b>/home/cpeel/public_html/c.branch/api-update-project/api/v1_projects.inc</b> on line <b>534</b><br />

message via Swagger/`curl`. (It may be something I'm doing wrong, but I thought I ought to mention it.)

cpeel commented:
Helps if I update the sandbox after rebasing the branch! Please try it now.

srjfoo commented:
> Helps if I update the sandbox after rebasing the branch! Please try it now.

I literally just did. Just before I saw your message notification. Success!
